ENH: clearer error msg for unequal categoricals in merge_asof (GH#26136)

chrish42 opened the pull request:
- [x] closes #26136 
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

chrish42 committed:
ENH: clearer error msg for unequal categoricals in merge_asof (GH#26136)

WillAyd commented on pandas/tests/reshape/merge/test_merge_asof.py:
```diff
@@ -1006,6 +1006,18 @@ def test_merge_datatype_error(self):
         with pytest.raises(MergeError, match=msg):
             merge_asof(left, right, on='a')
 
+    def test_merge_datatype_categorical_error(self):
```
Can you append `_raises` to the test name here?

chrish42 replied:
Sure. This was copied from `test_merge_datatype_error` (just above), though. Do I do that for that one too?

WillAyd replied:
Yea that should be fine to include here

WillAyd commented on pandas/tests/reshape/merge/test_merge_asof.py:
```diff
@@ -1006,6 +1006,18 @@ def test_merge_datatype_error(self):
             merge_asof(left, right, on='a')
 
+    def test_merge_datatype_categorical_error(self):
+        """ Tests merge datatype mismatch error """
```
Can remove docstring for test - we don't have these anywhere else

chrish42 replied:
Likewise. I do that for `test_merge_datatype_error` too?

chrish42 replied:
There are actually 4 other uses of docstrings in tests in this file. I convert them all to comments?

WillAyd replied:
I guess ignore for now but still remove this. Can have a follow up PR to make that module more consistent

WillAyd reviewed (changes requested):
Do we have a test for this that ensures this works when they are "equal"?

chrish42 commented:
Good question, actually. Yes, there is. It's `test_basic_categorical` (in the same file). But before I added the new test in this pull request, that was the only test that exercised the code paths of `merge_asof` for categoricals. I feel there should be more of those... That would be another pull request IMHO. though.

chrish42 committed:
Feedback from code review

WillAyd commented:
Thanks for clarifying. So this change does introduce unordered Categoricals which might be counter to what merge_asof is supposed to offer (the docstring explicitly states the join field must be ordered).

Do you you see any other instances of that method supporting unordered data? Wondering if we need to further require that Categoricals are ordered to use this method

jreback commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -37,6 +37,7 @@ Other Enhancements
 - :class:`RangeIndex` has gained :attr:`~RangeIndex.start`, :attr:`~RangeIndex.stop`, and :attr:`~RangeIndex.step` attributes (:issue:`25710`)
 - :class:`datetime.timezone` objects are now supported as arguments to timezone methods and constructors (:issue:`25065`)
 - :meth:`DataFrame.query` and :meth:`DataFrame.eval` now supports quoting column names with backticks to refer to names with spaces (:issue:`6508`)
+- :func:`merge_asof` now gives a clearer error message when merge keys are categoricals that are not equal (:issue:`26136`)
```
more clear error message

jreback commented on pandas/core/reshape/merge.py:
```diff
@@ -1446,10 +1446,18 @@ def _get_merge_keys(self):
-                                 "{rkdtype}, must be the same type"
-                                 .format(i=i, lkdtype=lk.dtype,
-                                         rkdtype=rk.dtype))
+                if (is_categorical_dtype(lk.dtype) and
```
so instead of this, i would instead just use ``repr(lk.dtype)`` and same for the right; this is a slight change for all dtypes, see if any tests break.

chrish42 replied:
What do you mean exactly? Using `repr(lk.dtype) == repr(rk.dtype)` instead of the `is_dtype_equal(...)` comparison? Or instead of the `is_categorical_dtype(...)`? Or both?

jreback replied:
neither use repr(lk.dtype) in the formatter

chrish42 replied:
Makes more sense. Done everywhere in `_get_merge_keys()`. Adjusted the tests that matched on the MergeError exception message to ignore the exact representation of the dtype.

chrish42 committed:
Tweak changelog message

chrish42 committed:
Work in progress for new test for unordered categoricals

chrish42 commented:
@WillAyd Indeed, the merge keys should be ordered, otherwise `merge_asof()` makes no sense. Currently, the merge fails later (in `_get_join_indexers()`) with a `ValueError`. But fun fact, there is no test currently in test_merge_asof.py. I've added a test that undordered categorical merge keys raises.

I've also added code to raise a MergeError (instead of a ValueError further down the road). But maybe adding this error check in in `_get_merge_keys()` is not the right place, because by then, the `by=` keys have been added to the list of merge keys.

Is there a reliable way to skip over the `by=` keys in this new test? Or should I put this test in another method? Thanks!

WillAyd commented:
Hmm that's unfortunate. If it makes things too complicated I'm not against looking that in more detail as a follow up if you can open an issue for it

chrish42 committed:
Revert "Work in progress for new test for unordered categoricals"

This reverts commit 71dcb80554d0870bb306f93416d8d1bbb17b66f1.

chrish42 committed:
Add comment to clarify why the test is not checking for ordered categories

chrish42 commented:
Cool. I've reverted the last change, and added a comment to explain what's happening currently. I've created #26249 for the other cleanups.

WillAyd reviewed (approved):
Thanks for opening an issue for follow ups. I'm OK with this as is with cleanups to follow. Over to @jreback

chrish42 committed:
Use repr(x.dtype) in MergeError messages

jreback commented:
thanks @chrish42 nice patch!
